File: src/main/java/frc/robot/commands/MoveArmPosition.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package frc.robot.commands;

import edu.wpi.first.math.geometry.Rotation2d;
import edu.wpi.first.wpilibj.DriverStation;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
import edu.wpi.first.wpilibj2.command.CommandBase;
import frc.robot.Constants.Arm;
import frc.robot.subsystems.ArmSubsystem;
import frc.robot.subsystems.ArmSubsystem.Positions;
import frc.robot.utils.MovingAverage;

/**
 * An example command that uses an example subsystem.
 */
public class MoveArmPosition extends CommandBase {

  private ArmSubsystem arm;
  private final Positions pos;
  private MotionLocation phase;
  private double currentAverage;

  private MovingAverage rotateMovingAverage;

  public enum MotionLocation {
    MovingIn,
    GoingToLocation,
    Extending
  };

  /**
   * bring the intake in or out
   * */
  public MoveArmPosition(ArmSubsystem arm, Positions pos) {
    this.arm = arm;
    this.pos = pos;

    this.rotateMovingAverage = new MovingAverage(Arm.rollingAverageLength);
    phase = MotionLocation.MovingIn;
    
    
    
    addRequirements(arm);
  }

  @Override
  public void initialize() {
    arm.setDesiredDistance(-1);
  }

  @Override
  public void execute() {
    DriverStation.reportWarning("arm extension auto: " + arm.getExtension(), false);
    DriverStation.reportWarning("arm rotation auto: " + arm.getRotation(),false);

    currentAverage = rotateMovingAverage.addValue(arm.getRotation());

    SmartDashboard.putNumber("current average of arm position", currentAverage);

    switch (phase) {
      case MovingIn:
        if (arm.getExtension() >= -2) {
          arm.setDesiredRotation(arm.getPostionAngle(pos));
          phase = MotionLocation.GoingToLocation;
        }
      break;
      case GoingToLocation:
        if (Math.abs(arm.getPostionAngle(pos) - currentAverage) < 4) {
          phase = MotionLocation.Extending;
          arm.setDesiredDistance(arm.getPostionExtension(pos));
        }
      break;
      case Extending:
      break;
    }
  }

  // Returns true when the command should end.
  @Override
  public boolean isFinished() {
    if (phase == MotionLocation.Extending
      && Math.abs(arm.getExtension() - arm.getPostionExtension(pos)) < 4) {
      return true;
    }
    return false;
  }
}

File: src/main/java/frc/robot/commands/MoveArmPosition.kt
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.
package frc.robot.commands

import edu.wpi.first.wpilibj.DriverStation
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard
import edu.wpi.first.wpilibj2.command.CommandBase
import frc.robot.Constants.Arm
import frc.robot.subsystems.ArmSubsystem
import frc.robot.subsystems.ArmSubsystem.Positions
import frc.robot.utils.MovingAverage
import kotlin.math.abs

/**
 * An example command that uses an example subsystem.
 */
class MoveArmPosition(private val arm: ArmSubsystem, private val pos: Positions) : CommandBase() {
    private var phase: MotionLocation
    private var currentAverage = 0.0
    private val rotateMovingAverage = MovingAverage(Arm.rollingAverageLength)

    enum class MotionLocation {
        MovingIn,
        GoingToLocation,
        Extending
    }

    /**
     * bring the intake in or out
     */
    init {
        phase = MotionLocation.MovingIn
        addRequirements(arm)
    }

    override fun initialize() {
        arm.desiredDistance = 0.0
    }

    override fun execute() {
        DriverStation.reportWarning("arm extension auto: " + arm.extension, false)
        DriverStation.reportWarning("arm rotation auto: " + arm.rotation, false)
        currentAverage = 360 - rotateMovingAverage.addValue(arm.rotation)
        SmartDashboard.putNumber("current average of arm position", currentAverage)
        when (phase) {
            MotionLocation.MovingIn -> if (arm.extension >= -0.2) {
                arm.desiredHeight = arm.getPostionAngle(pos)
                phase = MotionLocation.GoingToLocation
            }

            MotionLocation.GoingToLocation -> if (Math.abs(arm.getPostionAngle(pos) - currentAverage) < 4) {
                phase = MotionLocation.Extending
                arm.desiredDistance = arm.getPostionExtension(pos)
            }

            MotionLocation.Extending -> {}
        }
    }

    // Returns true when the command should end.
    override fun isFinished(): Boolean {
        return (phase == MotionLocation.Extending
                && abs(arm.extension + arm.getPostionExtension(pos)) < 0.04)
    }
}
